### ISSUE TITLE
aggregate plugin measurements into a single measurement

### ISSUE THREAD

**Issue by gotyaoi** [influxdata/telegraf#152]
It might make sense to aggregate the readings from a plugin into a single measurement, rather than having multiple measurements with matching timestamps. For example, the disk plugin currently sends multiple measurements (total, free, used, etc...), each containing a single field called value. This would be changed to a single measurement named disk with multiple fields (total, free, used, etc...) in it. The advantage of this would be that we could calculate derived values from such measurements in influxdb 0.9+. For example, the disk plugin currently has no option to calculate percentages and send those as well. If all the values were in one measurement, the percentages could be calculated on the fly, and the disk plugin would not have to support calculating and sending percentages.

**Comment by sparrc** [comment]
That's a very good point, I like this idea

/cc @beckettsean @toddboom what do you guys think? Any reason not to have a plugin's output contained within a single series?

**Comment by beckettsean** [comment]
@sparrc I do think it makes more sense to write into a single measurement with multiple fields, rather than decoupled measurements. It makes math on the results much easier.

The downside is that it means querying `disk_free` requires pulling all the points from disk, including the other unused fields. That's potentially a lot of unnecessary I/O if users are only graphing one value. Then again, all values can be trivially queried with a single select statement.

**Comment by skynet** [comment]
How about leaving it as is, with single fields, and adding another measurement with all fields combined.

**Comment by sparrc** [comment]
@beckettsean wouldn't it just be `SELECT disk_free FROM disk` rather than `SELECT * FROM disk_free`?

@skynet is there a reason that combining the fields into a single measurement won't work for you?

**Comment by beckettsean** [comment]
@sparrc `SELECT disk_free FROM disk` still has to pull the entire point from disk, including all field values. The DB cannot extract only the `disk_free` field. It's not a huge consideration but one thing to watch for after the switch.

Perhaps we can make it a configuration option? Some users want it all in one measurement, some want it in multiple measurements. Rather than doubling the data by writing to both let's give users the choice.

**Comment by sparrc** [comment]
Sounds fair, I guess I'm just worried that it might confuse and overly complex the documentation,

ie,

### Querying Data
To get CPU Idle percentage, use
```
SELECT cpu_percentageIdle FROM cpu
```
unless you set `separate_series = true`, then
```
SELECT value FROM cpu_percentageIdle
```

It's not a huge deal, I'm just not sure I understand the use-case of continuing to support the latter.

**Comment by beckettsean** [comment]
Good points, @sparrc. I, too, don't have much use for the multi-measurement setup. Maybe we should ask on the mailing list?

@pauldix want to weigh in?

**Comment by skynet** [comment]
@sparrc as long as we can select what's being stored in InfluxDB from the data generated by a plugin, multi-measurement setup is just fine.

**Comment by pauldix** [comment]
I'm fine with storing them all as different fields against a single measurement name. The storage engine that will be released in 0.9.5 won't actually have to read all the fields off disk when querying one specific field so that won't be an issue then. Each field will be stored in its own contiguous block of space.

For the field names, they should have underscore casing, and not include the measurement name as part of it. For example: `cpu` measurement `idle_time` as the field.

**Comment by sparrc** [comment]
:+1:, will fix all field names as part of this as well

**Comment by jpalay** [comment]
Awesome to see this is in the works! This will be super useful for us. I'm not sure how far you've gotten on this, but about a month ago I made a similar change to the CPU plugin on a branch of our fork of telegraf. See the last two commits of https://github.com/Asana/telegraf/tree/cpu_percentages.

I added a few methods to the Accumulator to make it easier to send measurements with multiple fields, and then changed the CPU plugin to use those new methods. Feel free to use as much or as little of this as you like!

**Comment by jpalay** [comment]
The branch also removes the cpu "busy" measurement, since when all the other metrics are in the sam measurement the "busy" time can be calculated in an influxdb query

**Comment by sparrc** [comment]
very cool, thanks @jpalay, that will be useful. Currently we're thinking that we'll wait until InfluxDB 0.9.5 to implement this across the board. That is because an influx query currently will get ALL fields within the same measurement, even if you only want one or two of them.

**Comment by marcogallotta** [comment]
@sparcc Any update as to whether you guys will be implementing this change across the board? I work with @jpalay, and we'd be happy to help change some of the plugins we use to aggregate the metrics if it's something that will be better supported as you suggest will be the case in 0.9.5.

**Comment by sparrc** [comment]
@marcogallotta update is still the same as above, we'll be waiting for the next version of InfluxDB (0.9.5) which is gated on the new storage engine.

The new storage engine will remove query penalties, such as all fields being retrieved from measurements at the same time, there will also be very significant storage compression improvements.

I'm not sure the exact timeframe on that, but I would suspect it's not going to be more than 2 weeks. I will start working on a branch to make that change soon, which I'll make public and builds will be available via CircleCI and from source

**Comment by ekini** [comment]
How about creating global option `separate_series = true` (I would call it separate_measurements), to make AddWithFields behave differently?
Then, for example, in `net` plugin we could use
```go
acc.AddWithFields(
    "", // empty suffix in this case
   map[string]interface{}{
       "bytes_sent": io.BytesSent,
        "bytes_recv": io.BytesRecv,
        "packets_sent": io.PacketsSent,
        "packets_recv": io.PacketsRecv,
        "err_in": io.Errin,
        "err_out": io.Errout,
        "drop_in": io.Dropin,
        "drop_out": io.Dropout,
        },
   tags)
````
instead of
````go
	acc.Add("bytes_sent", io.BytesSent, tags)
	acc.Add("bytes_recv", io.BytesRecv, tags)
	acc.Add("packets_sent", io.PacketsSent, tags)
	acc.Add("packets_recv", io.PacketsRecv, tags)
	acc.Add("err_in", io.Errin, tags)
	acc.Add("err_out", io.Errout, tags)
	acc.Add("drop_in", io.Dropin, tags)
	acc.Add("drop_out", io.Dropout, tags)
````
(which is quite repeatable).
This will accomplish three goals:
1. Let users of older influxdb use telegraf without penalty of storing multiple values in a measurement. Or if they just want multiple measurements.
2. Let users of influxdb 0.9.5 aggreate values to one measurement
3. Use the same code in plugins, regardless of influxdb version, as well as unify accumulator interface
4. Later rename function AddWithFields() to Add() (as the commented in the code)

**Comment by sparrc** [comment]
@ekini I just don't see a strong justification for having separate measurements anymore. It complicates the documentation significantly (see https://github.com/influxdb/telegraf/issues/152#issuecomment-136461809)

There is a reason that Telegraf is still in the `0.1.X` phase of releases, so I don't find backwards-compatability enough of a justification at this point.

**Comment by aderumier** [comment]
Hi, any news now that influx 0.9.6 has been released ?

**Comment by sparrc** [comment]
@aderumier the new storage engine is in 0.9.6, although it's not the default. I will be making breaking changes soon to Telegraf and cutting a 0.3.0 branch, I just wanted to first make sure we had all the bugfixes for 0.2.x in before doing that.

0.2.x will continue to be supported for a time

**Comment by skynet** [comment]
:+1: how soon can we expect a release?

**Comment by sparrc** [comment]
@skynet I'd like to have a "beta" 0.3 version ready by December 20th

**Comment by aderumier** [comment]
@sparrc Great :+1:  Waiting for Christmas :)

**Comment by padcom** [comment]
I'm a little confused regarding the change in connection with the `httpjson` and `exec` plugins. Currently when an endpoint returns value that value is then stored in series like so

```
exec_prefix-from-telegraf.conf_json-field
httpjson_prefix-from-telegraf.conf_json-field
```

I wonder what would it look like after the switch? Would it be `exec_prefix` or `exec.prefix` (does it make a difference?) or could we finally make that 100% configurable and just state what should be the series name without gluing it together with the plugin name?

My case here is that I use the exec plugin to collect data from a set of sensors. currently what is being pushed to the database is more less like so:

```
exec_sensors_pressure
exec_sensors_temperature
exec_sensors_humidity
```

I add there the `lat` and `lon` as tags and the name of the `location` as well. I think it'd be a lot nicer to store it in the database in the series `sensors` with 3 tags (`name`, `lat`, `lon`) and 3 values (`humidity`, `pressure`, `temperature`). That way querying the data is a breeze and it doesn't occupy as much space as all 3 separate series (if I'm not mistaken).

I think that it would look just like the ability to create external plugins without needing to submit them to the project which is great for closed-source projects.

Do you think that it is something that you could squeeze in into the work that you do on the aggregation?

**Comment by sparrc** [comment]
@padcom This is one of the goals of this change. I agree that there needs to be more measurements that are fields rather than their own measurements.

In 0.3.0 I'm going to be changing the way in which the `exec` and `httpjson` plugins are defined in the config file, as well as adding some more plugin options for redefining measurement names, so you'll be able to do something like this:

```
# exec collector 1
[[plugins.exec]]
  # the command to run
  command = "/usr/bin/collector1"
  name_override = "sensors" # default measurement name would be `exec`

# exec collector 2
[[plugins.exec]]
  # the command to run
  command = "/usr/bin/collector2"
  name_override = "other_things"
```

As you can see, you won't be defining separate `services` within the same plugin, each plugin will be a single service in of itself. Secondly, `name_override` is going to be a new argument available to _all_ plugins that users can use to change a measurement name of a particular plugin. (There will also be `name_prefix` and `name_suffix`)

Anyways, I think that covers what you suggested, let me know if there is something I'm missing there from your original question.

**Comment by padcom** [comment]
That sounds PERFECT! Can't wait to get it implemented in my setup :) Thanks for all your hard work. It's much appreciated!

**Comment by padcom** [comment]
Just one comment: since you pointed out that all plugins will benefit from the name_override option it seems that monitoring multiple Java services with JMX and Jolokia will finally make sense :) That will be awesome!

**Comment by sparrc** [comment]
FYI: a beta build of 0.3.0 is now available, see here: https://github.com/influxdb/telegraf#version-030-beta

The README on the [0.3.0 branch](https://github.com/influxdb/telegraf/tree/0.3.0) may also be helpful, as well as the [configuration guide](https://github.com/influxdb/telegraf/blob/0.3.0/CONFIGURATION.md).

**Comment by padcom** [comment]
I have tried the exec and httpjson plugins. Both, the command and an HTTP service return the same data:

### exec
```
{ "pressure": 99271, "temperature": 26.8 }
```

### httpjspn
```
* Rebuilt URL to: http://home.aplaline.com:32768/
* Hostname was NOT found in DNS cache
*   Trying 195.82.163.137...
* Connected to home.aplaline.com (195.82.163.137) port 32768 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: home.aplaline.com:32768
> Accept: */*
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Content-Type: application/json
< 
{ "pressure": 99264, "temperature": 26.9 }
* Closing connection 0
```
The `exec` plugin performs as advertised resulting in a series `sensors` with a tag `host` and 2 values for `pressure` and `temperature`. However the `httpjson` plugin returns with the following exception:

```
$ ./telegraf -config /etc/opt/telegraf/telegraf.conf -test
* Plugin: httpjson, Collection 1
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5691f8]

goroutine 36 [running]:
github.com/influxdb/telegraf/plugins/httpjson.(*HttpJson).sendRequest(0xc820181b20, 0xc82017ee01, 0x15, 0x0, 0x0, 0x0, 0x0)
	/home/telebuild/go/src/github.com/influxdb/telegraf/plugins/httpjson/httpjson.go:185 +0xd28
github.com/influxdb/telegraf/plugins/httpjson.(*HttpJson).gatherServer(0xc820181b20, 0x7f917edc26b0, 0xc82026e2c0, 0xc82017ee01, 0x15, 0x0, 0x0)
	/home/telebuild/go/src/github.com/influxdb/telegraf/plugins/httpjson/httpjson.go:122 +0x71
github.com/influxdb/telegraf/plugins/httpjson.(*HttpJson).Gather.func1(0xc820274a20, 0xc820181b20, 0x7f917edc26b0, 0xc82026e2c0, 0xc820190300, 0xc82017ee01, 0x15)
	/home/telebuild/go/src/github.com/influxdb/telegraf/plugins/httpjson/httpjson.go:89 +0x7e
created by github.com/influxdb/telegraf/plugins/httpjson.(*HttpJson).Gather
	/home/telebuild/go/src/github.com/influxdb/telegraf/plugins/httpjson/httpjson.go:92 +0x186

goroutine 1 [runnable]:
sync.runtime_Semacquire(0xc820274a2c)
	/usr/local/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc820274a20)
	/usr/local/go/src/sync/waitgroup.go:126 +0xb4
github.com/influxdb/telegraf/plugins/httpjson.(*HttpJson).Gather(0xc820181b20, 0x7f917edc26b0, 0xc82026e2c0, 0x0, 0x0)
	/home/telebuild/go/src/github.com/influxdb/telegraf/plugins/httpjson/httpjson.go:95 +0x1b3
github.com/influxdb/telegraf.(*Agent).Test(0xc8201780f8, 0x0, 0x0)
	/home/telebuild/go/src/github.com/influxdb/telegraf/agent.go:197 +0x49e
main.main()
	/home/telebuild/go/src/github.com/influxdb/telegraf/cmd/telegraf/telegraf.go:115 +0xe3c

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 20 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 35 [runnable]:
github.com/influxdb/telegraf.(*Agent).Test.func1(0xc8201902a0, 0xc820190240)
	/home/telebuild/go/src/github.com/influxdb/telegraf/agent.go:176
created by github.com/influxdb/telegraf.(*Agent).Test
	/home/telebuild/go/src/github.com/influxdb/telegraf/agent.go:185 +0xd6
```

**Comment by padcom** [comment]
The service http://home.aplaline.com:32768 is public facing - feel free to use it

**Comment by padcom** [comment]
Apart from the above I have tested the Apache plugin and the Docker plugin. They work like a charm!

**Comment by sparrc** [comment]
Thanks for the report @padcom, I'll look into that

**Comment by sparrc** [comment]
@padcom this has been fixed, you can now download 0.3.0-beta2 from the README.

Thanks again for the report!

**Comment by padcom** [comment]
Installed the latest, tested both (exec and httpjson) - all OK. Thank you!!

**Comment by padcom** [comment]
I have noticed a memory use reduction by about 400MB. Might not be related (changed from exec to httpjson plugin but also did more changes in the mean time) but it always is nice to see gauges go down :)

**Comment by sparrc** [comment]
closed by 0.10.0
